*: de-flake TestPaginatedBackupTenant and remove data race

irfansharif opened the pull request:
See individual commits; this PR squashes a data race in KV among other things.

---

This test readily fails under stress with span configs enabled (#73876).
Now that we split within the tenant keyspace for tenant tables, the
export requests generated for each BACKUP is liable to be retried if
concurrent with range splits. The splits causes KV to error out with
RangeKeyMismatchErrors, at which point the request is internally
retried. The test, which wants to assert on what ExportSpan requests are
intercepted, previously did not need to consider these retries/need to
de-dup -- it now does.

Release note: None

irfansharif committed:
backupccl: de-flake TestPaginatedBackupTenant

This test readily fails under stress with span configs enabled (#73876).
Now that we split within the tenant keyspace for tenant tables, the
export requests generated for each BACKUP is liable to be retried if
concurrent with range splits. The splits causes KV to error out with
RangeKeyMismatchErrors, at which point the request is internally
retried. The test, which wants to assert on what ExportSpan requests are
intercepted, previously did not need to consider these retries/need to
de-dup -- it now does.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74926)
<!-- Reviewable:end -->

irfansharif committed:
server: de-flake TestStatusAPICombinedTransactions

This test got a lot slower under the span configs infrastructure,
something we're investigating as part of #74919. Until then, prevent
this test from spoiling builds.

Release note: None

irfansharif committed:
kvserver: squash data race

Fixes #74932; we were reading from the replicas map without having
acquired the prerequisite lock.

Release note: None

tbg commented on pkg/ccl/backupccl/backup_test.go:
```diff
@@ -6816,10 +6818,12 @@ func TestPaginatedBackupTenant(t *testing.T) {
+					if _, found := exportRequestSpansSet[req]; found {
+						return nil // nothing to do
+					}
+					exportRequestSpansSet[req] = struct{}{}
```
Doesn't this want a mutex? Even if we "know" that there are no concurrent SSTs in this test, it would be better not to put in this assumption.

irfansharif replied:
Done.

tbg commented on pkg/kv/kvserver/store.go:
```diff
@@ -2336,6 +2336,9 @@ func (s *Store) onSpanConfigUpdate(ctx context.Context, updated roachpb.Span) {
 
 	now := s.cfg.Clock.NowAsClockTimestamp()
+
+	s.mu.RLock()
```
Aha!

tbg commented on pkg/server/status_test.go:
```diff
@@ -1478,6 +1478,7 @@ func TestStatusAPICombinedTransactions(t *testing.T) {
 	defer log.Scope(t).Close(t)
 
 	params, _ := tests.CreateTestServerParams()
+	params.Knobs.SpanConfig = &spanconfig.TestingKnobs{ManagerDisableJobCreation: true} // TODO(irfansharif): #74919.
```
Not sure if this is supposed to be merged?

irfansharif replied:
It is, yes. Investigating https://github.com/cockroachdb/cockroach/issues/74919 separately.

tbg reviewed (approved):
The "sqush race" commit needs to be backed out.

irfansharif commented:
bors r+

arulajmani reviewed (approved):
> Reviewed 3 of 3 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74926)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @irfansharif, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->
